feat: support network request aborting and refetching

amcgee opened the pull request:
This automatically aborts requests when the component is unmounted, the query changed, or a refetch initiated (new feature 👍).

This also adds `eslint-plugin-react-hooks` to eslint, allowing us to better check hooks.  **We should probably add these to `cli-style`!

This does not expose an imperative `abort` interface, though that could be added later if there's a need for it.  As it stands, I think it's better to let the component lifecycle handle the inflight requests, especially since we might not always have 1-to-1 query-to-request relationships.

We were missing dependencies for the `useEffect` in `useDataQuery`, which I've added here, so now the data will be re-requested if the query changes.

amcgee commented:
@varl is there a reason we aren't using **any** eslint plugins (most notable `eslint-plugin-react`) in `cli-style`?  Are we trying to keep it uber-generic?

varl commented:
> 
> 
> @varl is there a reason we aren't using **any** eslint plugins (most notable `eslint-plugin-react`) in `cli-style`? Are we trying to keep it uber-generic?

[Keeping it "JavaScript" generic](https://github.com/dhis2/notes/issues/52#issuecomment-501633608):

> The intended scope for d2-style is that JavaScript code standards should apply to all JavaScript code, not just React.

We could [add a group](https://github.com/dhis2/notes/issues/52#issuecomment-511428013) for React though, so you get a different ESLint configuration by doing e.g. `d2-style setup project/react`.

The "base" ESLint should be kept agnostic and only enforce vanilla JavaScript standards.

amcgee commented:
> Keeping it "JavaScript" generic

Cool cool, though so.  I think having a React group might make sense, but for now I've just added some eslint React rules here.

@varl thoughts this change and particularly whether or not to include an imperative `abort()` option in the hook output?

varl commented:
> 
> 
> > Keeping it "JavaScript" generic
> 
> Cool cool, though so. I think having a React group might make sense, but for now I've just added some eslint React rules here.

Agreed. The reason we don't have it yet is this:

> One thing to think about in that case is how to install the required dependencies.
>
> Do we want those in cli-style (can lead to bloat if we support a lot of custom projects)? Maybe the tool will tell you want deps you need to satisfy in your project instead.

varl commented:
> 
> @varl thoughts this change and particularly whether or not to include an imperative `abort()` option in the hook output?

Generally it would be nice to not have to think about it, but there might be scenarios where we do need to manually abort a request. Unless we actually need it now, my vote goes to only aborting when the component unmounts for now and adding it when we have use case we want to support.

amcgee commented:
> > @varl thoughts this change and particularly whether or not to include an imperative `abort()` option in the hook output?
> 
> Generally it would be nice to not have to think about it, but there might be scenarios where we do need to manually abort a request. Unless we actually need it now, my vote goes to only aborting when the component unmounts for now and adding it when we have use case we want to support.

Cool, I think I agree.  I'll leave it out for now, will be adding imperative refetch shortly which is probably more useful and has actually been requested

amcgee commented:
@varl updated to remove the imperative abort for now.  Also refactored the hook implementation to be cleaner and prepare for refetch

varl commented on services/data/src/hooks/useDataQuery.ts:
```diff
@@ -1,29 +1,52 @@
+import { ContextType } from '../types/Context'
 
-export const useDataQuery = (query: QueryMap): QueryState => {
+const reduceRepsonses = (responses: any[], names: string[]) =>
```
Typo: `reduceRepsonses`

amcgee commented:
There's a [known issue](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/33467) with duplicate typings in `react`/`react-dom` with `yarn`, fixed by adding `@types/**/node_modules` to `.yarnclean`

amcgee commented:
@varl thanks for the reviews!  Fixed the typo.

I went ahead and added the refetch feature here so that we don't need a second minor version bump, so I dismissed your approval.

I added an imperative `refetch` interface, cleaned up several sections, added tests, and upgraded the devDep on React/React-Dom to 16.9 so we don't have warnings flooding test output when testing async logic with `act`.  I've left the peerDep in the `runtime` package at `^16.8.6` though, which `16.9` satisfies.

janhenrikoverland commented on services/data/src/hooks/useDataQuery.ts:
```diff
@@ -1,29 +1,66 @@
 
-export const useDataQuery = (query: QueryMap): QueryState => {
+const reduceResponses = (responses: any[], names: string[]) =>
+    responses.reduce(
```
Consider mutating and returning the same object for performance reasons?

amcgee replied:
Thanks @janhenrikoverland !  Since this is a manually-constructed object with at most a few but usually 1 key (each key is a parallel data query) the performance implications should be negligible.  I'll change it though incase anyone gets really crazy with their query definitions (you'd need at least 100s to see performance degradation, I think)

amcgee commented:
@HendrikThePendric made a good point about AbortController in IE11 - we'll either need to polyfill it (not difficult) or guarantee that this will only be used in 2.33+

amcgee commented:
(Note: IE 11 also doesn't have `fetch`, so we're presumably polyfilling that already in cases when this is used for pre-2.33 apps/libs like Headerbar?  @varl?)

HendrikThePendric commented on services/data/src/hooks/useDataQuery.ts:
```diff
@@ -1,29 +1,62 @@
-    }, [])
-
-    return state
+        const controller = new AbortController()
```
IE11 lacks support for the `AbortController`. From DHIS 2 version 2.33 we will drop support for IE 11, so most apps using this hook won't suffer from this problem. However, there may be some exceptions:
- Apps that transitioned to feature toggling prior to 2.33: if at a later stage a developer decides to introduce the DataQuery component, this could cause problems (they can probably be avoided somehow)
- The new HeaderBar component: I believe this uses the DataQuery too, and it is being used in a lot of v2.32 apps. So this could be another potential source of problems.

My preferred solution would be to do the following:
- Keep the code as it is
- Add a section to the README (preferably fairly prominently displayed) mentioning:
   - That this lib relies on the `Fetch`, `AbortController` and `AbortSignal` APIs which are unsupported in IE11
   - If your app requires IE11 support, add a polyfill for all these APIs
   - Include some instructions on how to do so, i.e. linking to [whatwg-fetch](https://github.com/github/fetch) and [abortcontroller-polyfill](https://github.com/mo/abortcontroller-polyfill)

This each app can handle things as it sees fit.

varl commented:
@HendrikThePendric @amcgee We use a similar approach in ui-core where the App needs to provide polyfills: https://github.com/dhis2/ui-core#requires-polyfills

amcgee commented:
I’ll update the readme and merge tomorrow

Would be SO COOL to formally expose lib poly fill reqs to consuming apps!  One day...

amcgee commented:
Merging abort and refetch for a 1.5 release!
